plugins/state_history_plugin/include/eosio/state_history_plugin/state_history_log.hpp

```diff
@@ -188,13 +202,14 @@ class state_history_log {
       log.open(log_filename, std::ios_base::binary | std::ios_base::in | std::ios_base::out | std::ios_base::app);
       log.seekg(0, std::ios_base::end);
       uint64_t size = log.tellg();
-      if (size >= sizeof(state_history_log_header)) {
+      if (size >= state_history_log_header_serial_size) {
          state_history_log_header header;
          log.seekg(0);
-         log.read((char*)&header, sizeof(header));
-         EOS_ASSERT(header.version == 0 && sizeof(header) + header.payload_size + sizeof(uint64_t) <= size,
+         read_header(header, false);
+         EOS_ASSERT(is_ship(header.magic) && is_ship_supported_version(header.magic) &&
+                        state_history_log_header_serial_size + header.payload_size + sizeof(uint64_t) <= size,
```

Comment:
Last part of condition could be `log.tellg() + header.payload_size + sizeof(uint64_t) <= size`.